core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -106,13 +112,13 @@ class HadoopMapReduceCommitProtocol(
     val filename = getFilename(taskContext, ext)
 
     val stagingDir: Path = committer match {
-      case _ if dynamicPartitionOverwrite =>
-        assert(dir.isDefined,
-          "The dataset to be written must be partitioned when dynamicPartitionOverwrite is true.")
-        partitionPaths += dir.get
-        this.stagingDir
       // For FileOutputCommitter it has its own staging path called "work path".
       case f: FileOutputCommitter =>
+        if (dynamicPartitionOverwrite) {
+          assert(dir.isDefined,
+            "The dataset to be written must be partitioned when dynamicPartitionOverwrite is true.")
+          partitionPaths += dir.get
+        }
         new Path(Option(f.getWorkPath).map(_.toString).getOrElse(path))
```

Comment:
What about the case of `dynamicPartitionOverwrite=true` but `dir.isEmpty`?  IIUC, the workPath will be `/path/to/outputPath/.spark-staging-{jobId}/_temporary/{appAttemptId}/_temporary/{taskAttemptId}/` in this case. And it will be committed to `/path/to/outputPath/.spark-staging-{jobId}/` then. But it seems we don't move them to the `/path/to/outputPath/` in the end.

Reply:
IFAIK，`assert(dir.isDefined, ...)` already avoid this case, `dir.isDefined` means` !dir.isEmpty`, correct me if wrong :)

Reply:
oh, I see. My mistake.